### DISCUSSION TITLE
cannot get match graphql operations with msw & msw-storybook-addon

### DISCUSSION THREAD

**Discussion by z3nm0de** [mswjs/msw#1094] [Q&A] [answered]
I have been able to get a `rest.get` handler intercepting and working, but for some reason, no matter what, I cannot get the `graphql.query` with the matching `operation` name to intercept?  I keep getting:

I've printed out the handlers before allowing the query to fire, I've confirmed matching origin and operation name... but no idea why it's not matching up?

![image](https://user-images.githubusercontent.com/95314282/153517987-c6714bfe-9aac-4551-a43d-f695d17e766d.png)

```
    "msw": "^0.36.8",
    "msw-storybook-addon": "^1.6.0",
```

**Comment by z3nm0de** [comment]
Very odd, it seems to work with `axios` but fails with `fetch`.  

I copied the curls of the a good request and a bad request, they are identical, however chrome dev tools shows that the `fetch` one fails:

<img width="682" alt="image" src="https://user-images.githubusercontent.com/95314282/153527857-9474f254-6bf8-4590-9ff8-e6d782de34ed.png">


I ran into this using graphql codegen react-query plugin -- the default fetcher generates like this:

```ts
function fetcher<TData, TVariables>(endpoint: string, requestInit: RequestInit, query: string, variables?: TVariables) {
  return async (): Promise<TData> => {
    const res = await fetch(endpoint, {
      method: 'POST',
      ...requestInit,
      body: JSON.stringify({ query, variables }),
    });

    const json = await res.json();

    if (json.errors) {
      const { message } = json.errors[0];

      throw new Error(message);
    }

    return json.data;
  }
}
```

This actually fails ^ and I have no idea why its failing.  I would send `{ endpoint: 'https://mock-graphql/graphql' }` along with it.

When I wrote my own custom fetcher with `axios`, the request was intercepted correctly... any ideas?  Does this have to do with the default `fetch` running in the browser or some shim interfering or something?

**Comment by kettanaito** [comment, marked as answer]
Hey, @z3nm0de. 

It looks like the request you compose is not considered a valid GraphQL request by MSW. Based on your screenshots above, I think I see the issue:

- Your request has `Content-Type: text/plain` header. 

Please make sure you mark your request body type as `application/json` when performing GraphQL requests manually (via `fetch`):

```diff
    const res = await fetch(endpoint, {
      method: 'POST',
      ...requestInit,
+     headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ query, variables }),
    });
```

Without it, you're instructing your actual server to expect a plain text body. Let the server know you're sending a JSON. 

I suspect this doesn't happen with `axios` because it infers the `Content-Type` header somehow. It will happen with the plain `fetch` call unless you specify the request body type correctly.

**Reply by cooperjones** [reply]
Can't thank you enough for this answer, solved my problem after several hours of headache! I knew that my graphql mocks weren't being matched for some reason, but I wasn't able to find much info on how msw matches graph requests. Not sure if I may have just missed that in the documentation, but if it's not there, it may be beneficial to include additional info regarding how msw matches graph requests to help with debugging. I know I wouldn't have figured this out if it wasn't for this discussion post.

**Reply by kettanaito** [reply]
No worries. MSW often relies on implicit platform knowledge as going into detail on the semantics of REST and GraphQL requests is simply beyond its scope. And although you can technically send any `Content-Type` as a GraphQL query and leave it to your service to decipher, that's actually incorrect per GraphQL specification. In this regard, we ask you to comply with that because it's difficult to distinguish between GraphQL and REST-compatible requests as it is.
